from django.shortcuts import render, get_object_or_404
from .models import Module
from django.core import serializers
from django.http import HttpResponse
import string

LIST_OF_MODULES_TO_SEARCH = {
'abaqus' : ['finite element analysis', 'Engineering'],
'abinit' : ['Computational Chemistry and Materials Science', 'materials modeling', 'density functional theory', 'optimization', 'simulation'],
'ADF' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics', 'quantum mechanics', 'density functional theory', 'parallel binaries'],
'AFNI' : ['Magnetic Resonance Imaging (MRI)'],
'alphafold' : ['Computational Chemistry and Materials Science'],
'alsa-lib' : ['Visualization Libraries and Tools'],
'ambertools' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'ampl' : ['Programming'],
'anicalculator': ['Genomics', 'ANI', 'Genomics - classification tools'],
'ansys' : ['Engineering'],
'AnsysEM' : ['Engineering'],
'Ansys-Fluent' : ['Engineering'],
'ant' : ['Programming', 'java'],
'ants' : ['Magnetic Resonance Imaging (MRI)'],
'anvio' : ['Genomics', 'Genomics - analytical tools', 'Genomics utilities - visualization tools'],
'aria' : ['Programming'],
'armadillo' : ['Linear Algebra', 'Numerical Methods'], ## Linear Algebra
'ase' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'aspera' : ['Programming'],
'atk' : ['Computational Chemistry and Materials Science'],
'ATLAS' : ['Linear Algebra', 'Numerical Methods'], ## Linear Algebra 
'atom' : ['Text Editor'],
'awscli' : ['Programming'],
'bam-readcount' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - analytical tools'],
'bamtools' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - analytical tools'],
'baron' : ['Numerical Methods'], ## Non-linear optimization
'bart' : ['Magnetic Resonance Imaging (MRI)'],
'basespace' : ['Genomics', 'Genomics - analytical tools', 'cloud'],
'bazel' : ['Programming'],
'BBMap' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - RNA-seq'],
'bcbio_nextgen' : ['Genomics', 'python', 'Genomics - analytical tools'],
'BCBToolKit' : ['Genomics', 'python', 'Genomics - analytical tools'],
'bcftools' : ['Genomics', 'Genomics - variant calling/analysis', 'Genomics utlities - analytical pipelines'],
'bcl2fastq' : ['Genomics', 'Genomics - data conversion/transfer'],
'beagle-cpu' : ['Genomics'],
'bedops' : ['Genomics', 'Genomics - analytical tools'],
'bedtools' : ['Genomics', 'Genomics - analytical tools'],
'bismark' : ['Genomics', 'bisulfite mapping', 'methylation calling'],
'blas-lapack' : ['Linear Algebra', 'Numerical Methods'], ## Linear Algebra 
'blast' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - classification tools'],
'blatsuite' : ['Genomics', 'Genomics - sequence alignment'],
'blcr' : ['parallel', 'checkpointing', 'Programming'],
'boost' : ['Programming'],
'bowtie' : ['Genomics', 'Genomics - sequence alignment'],
'bowtie2' : ['Genomics', 'Genomics - sequence alignment'],
'bsmap' : ['Genomics', 'Genomics - analytical tools', 'bisulfite mapping', 'methylation calling'],
'bwa' : ['Genomics', 'Genomics - sequence alignment'],
'cairo' : ['Visualization Libraries and Tools'],
'casa' : ['Astrophysics'],
'cassi' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis'],
'cellranger' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - analytical tools', 'RNA-seq'],
'cellranger-atac' : ['Genomics', 'Genomics - analytical tools'],
'CfitsIO' : ['Files and IO'],
'cgal' : ['Numerical Methods'], ## computational geometry
'charm' : ['parallel'],
'charmpp' : ['parallel'],
'checkm' : ['Genomics', 'quality assessment'],
'chemaxon_marvin' : ['Computational Chemistry and Materials Science'],
'chemaxon_structure_rep' : ['Computational Chemistry and Materials Science'],
'chimera' : ['Computational Chemistry and Materials Science', 'Visualization Libraries and Tools'],
'clang' : ['Programming'],
'cmake' : ['Programming'],
'cmdstan' : ['Programming'],
'connectome_workbench' : ['Magnetic Resonance Imaging (MRI)'],
'copasul' : ['Visualization Libraries and Tools'],
'cp2k' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics', 'quantum mechanics'],
'cplex' : ['Numerical Methods'], ## Optimizer
'CrisprCasFinder' : ['Genomics', 'Genomics - analytical tools'],
'cuda' : ['Programming'],
'cufflinks' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq'],
'curl' : ['Programming'],
'cytoscape' : ['Genomics', 'Genomics - analytical tools'],
'db' : ['Files and IO'],
'ddd' : ['Visualization Libraries and Tools'],
'deeptools' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq', 'ChIP-seq', 'MNase-seq'],
'diamond' : ['Genomics', 'Genomics - sequence alignment'],
'diffreps' : ['Genomics', 'Genomics - analytical tools', 'ChIP-seq'],
'dynare' : ['Economics'],
'dyninst' : ['Programming'],
'eclipse' : ['Programming', 'java'],
'eigen' : ['Linear Algebra', 'Numerical Methods'], ## Linear Algebra
'elfutils' : ['Files and IO'],
'emacs' : ['Text Editor'],
'epacts' : ['Genomics', 'Genomics - analytical tools'],
'espresso' : ['Computational Chemistry and Materials Science'],
'expansionhunter' : ['Genomics', 'Genomics - analytical tools'],
'expat' : ['Files and IO'],
'fastani' : ['Genomics', 'ANI'],
'fastqc' : ['Genomics', 'quality assessment'],
'fasttree' : ['Genomics', 'Genomics - taxonomic/phylogenetic tools'],
'FastTreeMP' : ['Genomics', 'Genomics - taxonomic/phylogenetic tools', 'parallel'],
'fenics27' : ['Programmings', 'finite element analysis', 'Engineering'],
'ffmpeg' : ['Files and IO', 'Visualization Libraries and Tools'],
'fftw' : ['Fourier Transform'], ## Fourier Transform
'fMRIPrep' : ['Magnetic Resonance Imaging (MRI)'],
'freesurfer' : ['Magnetic Resonance Imaging (MRI)'],
'freetype' : ['Programming'],
'fsl' : ['Magnetic Resonance Imaging (MRI)'],
'gamess' : ['Computational Chemistry and Materials Science'],
'gap' : ['Numerical Methods'], ## Computational Discrete Algebra
'gatk' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis', 'Genomics - analytical tools', 'Genomics - data conversion/transfer'],
'gauss' : ['Programming'],
'Gaussian' : ['Computational Chemistry and Materials Science'],
'Gaussian16' : ['Computational Chemistry and Materials Science'],
'gcc' : ['Programming'],
'gcloud' : ['Programming'],
'gdal' : ['Files and IO'],
'gdb' : ['Programming'],
'gdbm' : ['Files and IO'],
'gdc-client' : ['Genomics', 'Genomics - data conversion/transfer'],
'gdk-pixbuf' : ['Visualization Libraries and Tools'],
'geany' : ['Text Editor'],
'GeneTorrent' : ['Genomics', 'Genomics - data conversion/transfer'],
'GenomeBrowse' : ['Genomics', 'visualization tools'],
'geos' : ['Programming'],
'gettext' : ['Files and IO'],
'gffcompare' : ['Genomics', 'Genomics - RNA-seq', ''],
'gffread' : ['Genomics', 'Genomics - data conversion/transfer'],
'ghc' : ['Programming'],
'ghostscript' : ['Programming'],
'git' : ['Programming', 'version control software'],
'gnumeric' : ['Visualization Libraries and Tools'],
'gnuplot' : ['Visualization Libraries and Tools'],
'go' : ['Programming'],
'gold' : ['Genomics'],
'GotoBlas' : ['Linear Algebra', 'Numerical Methods'], ## Linear Algebra
'gpaw' : ['Computational Chemistry and Materials Science', 'density functional theory'],
'grace' : ['Visualization Libraries and Tools'],
'gradle' : ['Programming', 'IDE'],
'graphlib' : ['Visualization Libraries and Tools'],
'graphviz' : ['Visualization Libraries and Tools'],
'GRASS' : ['Visualization Libraries and Tools'],
'gromacs' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'gsl' : ['Linear Algebra', 'Numerical Methods', 'Programming'], ## Linear Algebra
'gstreamer' : ['Files and IO'],
'gtk+' : ['Visualization Libraries and Tools'],
'gtool' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis', 'Genomics - analytical tools'],
'guppy' : ['Genomics', 'visualization tools'],
'gurobi' : ['Numerical Methods'], ## Optimizer
'hdf5' : ['Files and IO'],
'hic-pro' : ['Genomics', 'Genomics - analytical tools', 'Hi-C'],
'hisat2' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - RNA-seq'],
'hmmer' : ['Genomics', 'Genomics - classification tools', 'Genomics - sequence alignment'],
'homer' : ['Genomics'],
'htslib' : ['Genomics', 'Genomics - sequence alignment'],
'hypre' : ['Numerical Methods'], ## solving linear systems of equations, optimizer
'icu4c' : ['Programming', 'java'],
'idba' : ['Genomics', 'Genomics - analytical tools', 'Genomics utilities - sequence assembler'],
'igv' : ['Genomics', 'visualization tools'],
'ImageJ' : ['Magnetic Resonance Imaging (MRI)', 'Visualization Libraries and Tools'],
'ImageMagick' : ['Visualization Libraries and Tools'],
'impute2' : ['Genomics', 'Genomics - analytical tools'],
'imsl' : ['Linear Algebra', 'Numerical Methods'],
'indelocator' : ['Genomics', 'Genomics - variant calling/analysis', 'Genomics - analytical tools'],
'intel' : ['Programming'],
'intel-mkl' : ['Linear Algebra', 'Numerical Methods', 'Fourier Transform'], ## Linear Algebra, Fourier Transform
'intltool' : ['Files and IO'],
'ipopt' : ['Numerical Methods'], ## non-linear optimization
'iread' : [],
'jags' : ['statistical computing'],
'java' : ['Programming'],
'jellyfish' : ['Genomics'],
'jq' : ['Files and IO'],
'juicer' : ['Genomics'],
'julia' : ['Programming'],
'kallisto' : ['Genomics', 'Genomics - RNA-seq'],
'kentUtils' : ['Genomics', 'Genomics - analytical tools', 'Genomics utilities - visualization tools'],
'kim-api' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'knitro' : ['Numerical Methods'], ## non-linear optimization
'kraken' : ['Genomics', 'Genomics - taxonomic/phylogenetic tools'],
'krona' : ['Genomics', 'visualization tools'],
'kSNP' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis'],
'lammps' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'latte' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'lazarus' : ['Programming', 'IDE'],
'leptonica' : ['Visualization Libraries and Tools'],
'lftp' : ['Files and IO'],
'libreoffice' : ['Visualization Libraries and Tools'],
'liggghts-public' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'lp_solve' : ['Numerical Methods'], ## linear solver
'lumerical' : ['programmining language', 'Computational Chemistry and Materials Science', 'FDTD'],
'MACS2' : ['Genomics', 'Genomics - analytical tools', 'ChIP-seq'],
'mafft' : ['Genomics', 'Genomics - sequence alignment'],
'mageck' : ['Genomics', 'Genomics - analytical tools'],
'mageck_vispr' : ['Genomics', 'Genomics - analytical tools', 'visualization tools'],
'makedepf90' : ['Programming'],
'manta' : ['Files and IO'],
'mathematica' : ['Programming'],
'matlab' : ['Programming'],
'matlabstan' : ['Numerical Methods'], ## Bayesian Inference
'meep' : ['Computational Chemistry and Materials Science', 'FDTD'],
'megahit' : ['Genomics', 'Genomics - assembly tools'],
'meme' : ['Genomics', 'Genomics - analytical tools', 'Genomics - analytical tools'],
'mercurial' : ['Programming', 'version control software'],
'metabat' : ['Genomics', 'Genomics - analytical tools'],
'metal' : ['Genomics', 'Genomics - analytical tools', ''],
'metamap' : ['Files and IO'],
'metaxa2' : ['Genomics', 'Genomics - taxonomic/phylogenetic tools'],
'metis' : ['Genomics', 'annotation tools', 'Numerical Methods'],
'mlpack' : ['Numerical Methods'],
'mono' : ['Programming'],
'motif' : ['Visualization Libraries and Tools'],
'mpb' : ['Computational Chemistry and Materials Science', 'Numerical Methods'],
'mpi' : ['parallel', 'Programming'],
'mpi4py' : ['parallel'],
'mricrogl' : ['Magnetic Resonance Imaging (MRI)'],
'mrtrix3' : ['Magnetic Resonance Imaging (MRI)'],
'multiqc' : ['Genomics', 'quality assessment'],
'mumps' : ['Programming'],
'muscle' : ['Genomics', 'Genomics - sequence alignment'],
'mutect' : ['Genomics', 'Genomics - analytical tools', 'Genomics - variant calling/analysis'],
'mysql' : ['Files and IO'],
'namd' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'nano' : ['Text Editor'],
'ncl' : ['Programming', 'Visualization Libraries and Tools'],
'nco' : ['Files and IO'],
'ncurses' : ['Visualization Libraries and Tools'],
'ndiff' : ['Files and IO'],
'netCDF' : ['Files and IO'],
'netcdf-c' : ['Files and IO'],
'netcdf-fortran' : ['Files and IO'],
'netlib-scalapack' : ['Linear Algebra', 'Numerical Methods'],
'netlogo' : ['Programming'],
'ngsplot' : ['Genomics', 'visualization tools'],
'nlopt' : ['Numerical Methods'], ## nonlinear optimization
'novocraft' : ['Genomics', 'Genomics - analytical tools'],
'numpy' : ['Python'],
'nwchem' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics', 'quantum mechanics'],
'ocaml' : ['Programming'],
'octave' : ['Programming'],
'oncotator' : ['Genomics', 'annotation tools', 'Genomics - variant calling/analysis'],
'openbabel' : ['Computational Chemistry and Materials Science'],
'openblas' : ['Linear Algebra', 'Numerical Methods'],
'OpenBugs' : ['statistical computing'],
'opencl' : ['Programming'],
'openmd' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'orca' : ['Computational Chemistry and Materials Science'],
'oxconsole' : ['IDE'],
'oxedit' : ['Text Editor'],
'p7zip' : ['Files and IO'],
'pandoc' : ['Files and IO'],
'papi' : ['Programming'],
'parallel' : ['parallel'],
'paraview' : ['Visualization Libraries and Tools'],
'parmetis' : ['Parallel', 'Numerical Methods'], ## Parallel Graph Partitioning and Fill-reducing Matrix Ordering.
'parsec' : ['Computational Chemistry and Materials Science'],
'pBWA' : ['Genomics', 'Genomics - sequence alignment', 'parallel'],
'pdflib-lite' : ['Files and IO'],
'peer' : [],
'penncnv' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis'],
'perl' : ['Programming'],
'petsc' : ['Numerical Methods'], ## The Portable, Extensible Toolkit for Scientific Computation, is a suite of data structures and routines developed by Argonne National Laboratory for the scalable solution of scientific applications modeled by partial differential equations
'pexsi' : ['Computational Chemistry and Materials Science', 'density functional theory'],
'phylocsf' : ['Genomics', 'Genomics - RNA-seq'],
'picard' : ['Genomics', 'Genomics - analytical tools', 'Genomics - variant calling/analysis', 'Genomics - sequence alignment'],
'pigz' : ['Files and IO'],
'pindel' : ['Genomics', 'Genomics - variant calling/analysis', 'Genomics - variant calling/analysis', 'Genomics - analytical tools'],
'plink' : ['Genomics', 'Genomics - analytical tools', 'Genomics utilities - visualization tools'],
'plumed' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'poppler' : ['Files and IO'],
'povray' : ['Visualization Libraries and Tools'],
'pplacer' : ['Genomics', 'Genomics - analytical tools', 'Genomics utilities - visualization tools'],
'prada' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq'],
'prodigal' : ['Genomics', 'annotation tools'],
'proj' : ['Programming'],
'protobuf' : ['Files and IO'],
'PRSice' : ['Genomics', 'Genomics - analytical tools'],
'python' : ['Programming'],
'python-anaconda3' : ['Programming'],
'python-miniconda3' : ['Programming'],
'qchem' : ['Computational Chemistry and Materials Science'],
'qctool' : ['Genomics', 'Genomics - variant calling/analysis', 'Genomics utilities - quality assessment'],
'qe' : ['Computational Chemistry and Materials Science'],
'qhull' : ['Numerical Methods'], ## computational geometry
'qiime2' : ['Genomics', 'Genomics - analytical tools', 'Genomics - taxonomic/phylogenetic tools'],
'qt' : ['Visualization Libraries and Tools'],
'qualimap' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - RNA-seq', 'ChIP-seq'],
'quantum-espresso' : ['Computational Chemistry and Materials Science'],
'R' : ['Programming', 'statistical computing'],
'raxml' : ['Genomics', 'Genomics - taxonomic/phylogenetic tools'],
'rfmix' : ['Genomics', 'Genomics - analytical tools', 'Genomics - taxonomic/phylogenetic tools'],
'rmats2sashimiplot' : ['Genomics', 'visualization tools', 'Genomics - RNA-seq'],
'rmats_turbo' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq'],
'rsem' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq'],
'ruby' : ['Programming'],
'rufus' : ['Genomics', 'Genomics - analytical tools', 'Genomics - variant calling/analysis'],
'sage' : ['Genomics', 'Genomics - analytical tools', 'epidemiology tools'],
'sailfish' : ['Genomics', 'Genomics - analytical tools', 'Genomics - RNA-seq'],
'salmon' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment'],
'samtools' : ['Genomics', 'Genomics - sequence alignment'],
'sas' : ['Programming'],
'scala' : ['Programming'],
'scalapack' : ['Linear Algebra', 'Numerical Methods'],
'scallop' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment'],
'schrodinger' : ['Material Science', 'Engineering'],
'scons' : ['Programming'],
'sdpt3' : ['MATLAB'], # optimization
'seqkit' : ['Genomics', 'Genomics - analytical tools'],
'seqtk' : ['Genomics', 'Genomics - analytical tools'],
'shapeit' : ['Genomics', 'Genomics - analytical tools', 'SNPs', 'Genomics - variant calling/analysis'],
'singularity' : ['Programming'],
'sirius' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'slepc' : ['Numerical Methods'], ## solution of large scale sparse eigenvalue problems on parallel computers.
'snakemake' : ['Programming'],
'snpEff' : ['Genomics', 'Genomics - sequence alignment', 'SNPs', 'Genomics - variant calling/analysis'],
'snptest' : ['Genomics', 'SNPs', 'Genomics - variant calling/analysis'],
'soapdenovo2' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - assembly tools'],
'SolexaQA++' : ['Genomics', 'quality assessment'],
'spades' : ['Genomics', 'Genomics - assembly tools'],
'spark' : ['Files and IO'],
'specfem3d_globe' : [],
'speedseq' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - assembly tools', 'SNPs', 'Genomics - variant calling/analysis'],
'spfft' : ['Numerical Methods', 'Fourier Transform'],
'spglib' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics'],
'spla' : ['Numerical Methods', 'Linear Algebra'],
'spm' : ['Magnetic Resonance Imaging (MRI)'],
'sqlite' : ['Files and IO'],
'sratoolkit' : ['Genomics', 'Genomics - data conversion/transfer'],
'STAR' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment'],
'STAR-Fusion' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment'],
'stat' : ['Numerical Methods'],
'stata' : ['Programming'],
'stattransfer' : ['Files and IO'],
'strelka' : ['Genomics', 'Genomics - variant calling/analysis'],
'stress' : ['Programming'],
'stringtie' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment'],
'subread' : ['Genomics', 'Genomics - RNA-seq', 'Genomics - sequence alignment', 'Genomics - variant calling/analysis'],
'SuiteSparse' : ['Numerical Methods'],
'superlu' : ['Numerical Methods'],
'superlu-dist' : ['Numerical Methods'],
'suppa' : ['Genomics', 'Genomics - RNA-seq'],
'swig' : ['Programming'],
'tar' : ['Files and IO'],
'tau' : ['Programming'],
'tcl' : ['Programming'],
'telseq' : ['Genomics', 'Genomics - analytical tools'],
'tesseract' : [],
'texlive' : ['Programming'],
'tickwrite' : ['Economics'],
'tig' : ['Programming', 'version control software', 'git'],
'tmux' : ['Programming'],
'tophat' : ['Genomics', 'Genomics - RNA-seq'],
'Tramonto' : ['Material Science', 'Biology'],
'TrimGalore' : ['Genomics', 'Genomics - analytical tools'],
'trimmomatic' : ['Genomics', 'Genomics - analytical tools'],
'uchime' : ['Genomics', 'quality assessment'],
'udunits2' : ['Programming'],
'utilities' : ['Programming', 'Programming'],
'valgrind' : ['Programming'],
'vasp' : ['Computational Chemistry and Materials Science'],
'vcftools' : ['Genomics', 'Genomics - variant calling/analysis', 'Genomics - analytical tools'],
'velvet' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - assembly tools'],
'vesta' : ['Material Science', 'Visualization Libraries and Tools'],
'vim' : ['Text Editor'],
'visit' : ['Visualization Libraries and Tools'],
'vmd' : ['Computational Chemistry and Materials Science', 'Molecular Dynamics', 'Visualization Libraries and Tools'],
'voropp' : ['Numerical Methods'],
'vtk' : ['Visualization Libraries and Tools'],
'wannier' : ['Computational Chemistry and Materials Science'],
'wine' : ['Programming'],
'wxwidgets' : ['Visualization Libraries and Tools'],
'xander_assembler' : ['Genomics', 'Genomics - sequence alignment', 'Genomics - assembly tools'],
'xcb-proto' : ['Programming'],
'xmlsec' : ['Files and IO'],
'xpdf' : ['Files and IO'],
'xz' : ['Files and IO'],
'yaml-cpp' : ['Files and IO'],
'yasm' : ['Programming'],
'zstd' : ['Files and IO'],
}

LIST_OF_MODULES_NOT_TO_SEARCH = [
'firefox',
'opera',
'Gaussian.E01',
'MS-Ropen',
'at-spi2-atk',
'at-spi2-core',
'bwrap',
'cloog',
'glib',
'glibmm',
'gmp',
'hwloc',
'isl',
'launchmon',
'libbsd',
'libdwarf',
'libepoxy',
'libffi',
'libgdiplus',
'libglvnd',
'libiconv',
'libjpeg',
'libjpeg-turbo',
'libpng',
'libsvm',
'libszip',
'libtiff',
'libtool',
'libvdwxc',
'libwayland-cursor',
'libxc',
'libxcb',
'libxml2',
'libxsmm',
'mpc',
'mpfr',
'mrnet',
'numactl',
'osu-micro-benchmarks',
'pbs-drmaa',
'pcre',
'pkg-config',
'readline',
'root',
'slurm-drmaa',
'zlib',
]

# Create your views here.
letters_a_to_z = [ i for i in string.ascii_uppercase]

def index(request):
    modules = Module.objects.exclude(name__in=LIST_OF_MODULES_NOT_TO_SEARCH)
    all_keywords = []
    for module in modules:
        if module.primary_keywords is not None:
            for keyword in module.primary_keywords:
                all_keywords.append(keyword)
    all_keywords = set(sorted(all_keywords))
    
    return render(request, 'modules/index.html', {"modules" : modules, "all_keywords" : all_keywords, 'letters_a_to_z' : letters_a_to_z})

def module_json(request):
    modules = Module.objects.exclude(name__in=LIST_OF_MODULES_NOT_TO_SEARCH)
    modules_list = serializers.serialize('json', modules)
    return HttpResponse(modules_list, content_type="text/json-comment-filtered")
